include/mbedtls/ssl.h

```diff
@@ -740,43 +740,64 @@ mbedtls_ssl_states;
 
 #if defined(MBEDTLS_SSL_EARLY_DATA) && defined(MBEDTLS_SSL_CLI_C)
 typedef enum {
-/*
- * The client has not sent the first ClientHello yet, it is unknown if the
- * client will send an early data indication extension or not.
- */
-    MBEDTLS_SSL_EARLY_DATA_STATUS_UNKNOWN,
-
 /*
  * See documentation of mbedtls_ssl_get_early_data_status().
  */
-    MBEDTLS_SSL_EARLY_DATA_STATUS_NOT_SENT,
+    MBEDTLS_SSL_EARLY_DATA_STATUS_NO_IND_SENT,
     MBEDTLS_SSL_EARLY_DATA_STATUS_ACCEPTED,
     MBEDTLS_SSL_EARLY_DATA_STATUS_REJECTED,
+} mbedtls_ssl_early_data_status;
+
+typedef enum {
+/*
+ * The client has not sent the first ClientHello yet, the negotiation of early
+ * data has not started yet.
+ */
+    MBEDTLS_SSL_EARLY_DATA_STATE_IDLE,
+
+/*
+ * In its ClientHello, the client has not included an early data indication
+ * extension.
+ */
+    MBEDTLS_SSL_EARLY_DATA_STATE_NO_IND_SENT,
 
 /*
  * The client has sent an early data indication extension in its first
  * ClientHello, it has not received the response (ServerHello or
  * HelloRetryRequest) from the server yet. The transform to protect early data
- * is not set and early data cannot be sent yet.
+ * is not set either as for middlebox compatibility a dummy CCs may have to be
+ * sent in clear. Early data cannot be sent to the server yet.
  */
-    MBEDTLS_SSL_EARLY_DATA_STATUS_SENT,
+    MBEDTLS_SSL_EARLY_DATA_STATE_IND_SENT,
 
 /*
  * The client has sent an early data indication extension in its first
  * ClientHello, it has not received the response (ServerHello or
  * HelloRetryRequest) from the server yet. The transform to protect early data
  * has been set and early data can be written now.
  */
-    MBEDTLS_SSL_EARLY_DATA_STATUS_CAN_WRITE,
+    MBEDTLS_SSL_EARLY_DATA_STATE_CAN_WRITE,
+
+/*
+ * The client has indicated the use of early data and the server has accepted
+ * it.
+ */
+    MBEDTLS_SSL_EARLY_DATA_STATE_ACCEPTED,
+
+/*
+ * The client has indicated the use of early data but the server has rejected
+ * it.
+ */
+    MBEDTLS_SSL_EARLY_DATA_STATE_REJECTED,
```

Comment:
(comment) I can appreciate that `state` makes a lot of sense, but now we have both `MBEDTLS_SSL_EARLY_DATA_STATUS_xxx` and `MBEDTLS_SSL_EARLY_DATA_STATE_xxx` which are hard to distinguish (one or two letters at the end of a word in the middle of the identifier).

I don't have any good suggestions about this (yet?) but wonder if other readers are finding this an issue too?

Reply:
I see you point, I think the problem is not too bad `STATUS_xxx` code are only used in a single place: `mbedtls_ssl_get_early_data_status()`, and all TLS code except for that particular function will use the `STATE_xxx` ones, so a confusion seems rather unlikely.

If I could see an easy way to better distinguish without introducing other issues (like, one could use `0RTT` and the other `EARLY_DATA` but IMO that would make things worse as it would break the nice consistency we have so far as we use "early data" everywhere), sure. But see I'm not seeing one, I'm OK with the state of things.

Reply:
I was going to add "and since `STATE_xxx` is internal we can always rename it later if this is an issue and/or we think of a better name" but then [I realised it's actually public...](https://github.com/Mbed-TLS/mbedtls/pull/8884/files#r1519491941)

Reply:
I've been through this state/status issue when implementing. I started with having only enumeration but it has its drawbacks as stated during the code review of #8760. For the same reason as mentioned by @mpg above, I feel it is not a too big maintenance issue to have state and status macros: status macros are only use at a specific place.

Reply:
I have not made any change related to that yet. Though about `MBEDTLS_SSL_EDATA_STATE_xxx` or `MBEDTLS_SSL_EDATA_STA_xxx` to differentiate more from `MBEDTLS_SSL_EARLY_DATA_STATUS_xxx` but not really convinced about it.